### ISSUE TITLE
need return nonce in GetMeasurement and Challenge to the caller

### ISSUE THREAD

**Issue by jyao1** [DMTF/libspdm#137]
The caller may need evaluate the nonce to see if there is any duplication.

**Comment by steven-bellock** [comment]
Is this to detect a faulty random number generator?

**Comment by jyao1** [comment]
Right. The detection is out of scope of libspdm. So we need provide a way to let caller detect, if the caller wants to.

**Comment by steven-bellock** [comment]
Right, it is out of scope of libspdm but I think the cryptography module should handle it, not the caller. For example FIPS 140 says

> Continuous random number generator test. If a cryptographic module employs Approved or nonApproved RNGs in an >Approved mode of operation, the module shall perform the following continuous
> random number generator test on each RNG that tests for failure to a constant value.
> 1. If each call to a RNG produces blocks of n bits (where n > 15), the first n-bit block generated
> after power-up, initialization, or reset shall not be used, but shall be saved for comparison with
> the next n-bit block to be generated. Each subsequent generation of an n-bit block shall be
> compared with the previously generated block. The test shall fail if any two compared n-bit
> blocks are equal.
> 2. If each call to a RNG produces fewer than 16 bits, the first n bits generated after power-up,
> initialization, or reset (for some n > 15) shall not be used, but shall be saved for comparison with
> the next n generated bits. Each subsequent generation of n bits shall be compared with the
> previously generated n bits. The test fails if any two compared n-bit sequences are equal.

**Comment by jyao1** [comment]
This is more like attestation. 
The cryptography module need guarantee. And the remote verifier may check:
1) If the nonce from device is random.
2) If the nonce from local environment is random.

**Comment by steven-bellock** [comment]
Two comments.

- From the verifier's point of view the responder's nonce is immaterial to the overall security of Measurements and Challenge. The responder's nonce is meant to protect the responder and, with the cryptography algorithms that SPDM supports, probably contributes very little to that protection.
- This is part of a larger topic on the interplay between verifier and requester and where the verifier and requester reside within the overall system, and how libspdm can support the different configurations. I can raise that as a topic at the next meeting.

**Comment by steven-bellock** [comment]
@raghuncstate also has opinions on this. :)

**Comment by raghuncstate** [comment]
https://github.com/DMTF/Security-TF/issues/1438 discussed the issue. Steven's first point is correct. The responder nonce makes very little difference from a security point of view.

What is the goal of this issue? Is it to test that a number is random? proving that something is random is non-trivial and more so should not matter in the context of nonces. so long as the producer of hte nonce knows that it is random (or really the number is used only once, so not random, strictly speaking), the protocol security is not affected..so i'm not sure i understand what "The caller may need evaluate the nonce to see if there is any duplication." means..

**Comment by jyao1** [comment]
This is requested explicitly by the multiple customers.

The use case is: The verifier can cache all reported nonce and compare them in next round report. The goal is not to test if they are random (I don't know how to achieve that). The goal is to see if the nonce is duplicated with previous one. The verifier may be local or remote.

Device can claim it supports real nonce or not (by using different nonce or same nonce). The verifier owns the verification policy and can decide to accept the device or not. They are two different things.

**Comment by steven-bellock** [comment]
Alternatively the verifier could generate and provide the nonce to the requester.

**Comment by jyao1** [comment]
> Alternatively the verifier could generate and provide the nonce to the requester.

Agree. That is also possible.

**Comment by steven-bellock** [comment]
spdm_challenge_ex can be modified to take the nonce value as an input.
